rclcpp/src/rclcpp/executors/static_single_threaded_executor.cpp

```diff
@@ -0,0 +1,324 @@
+          if(waitable){
+            // If any waitable is found, push it in the exec_list struct
+            exec_list.waitable.push_back(waitable);
+            exec_list.number_of_waitables++;
+          }
+            return false;
+        });
+    }
+  }
+}
+
+void
+StaticSingleThreadedExecutor::get_executable_list(
+  ExecutableList & executable_list, std::chrono::nanoseconds timeout)
+{
+  // prepare the wait_set
+  prepare_wait_set();
+
+  // add handles to the wait_set and wait for work
+  refresh_wait_set(timeout);
+
+  // Get all the timers
+  get_timer_list(executable_list);
+
+  // Get all the subscribers
+  get_subscription_list(executable_list);
+
+  // Get all the services
+  get_service_list(executable_list);
+
+  // Get all the clients
+  get_client_list(executable_list);
+
+  // Get all the waitables
+  get_waitable_list(executable_list);
+}
+
+void
+StaticSingleThreadedExecutor::execute_ready_executables(
+  ExecutableList & exec_list, std::chrono::nanoseconds timeout)
+{
+  refresh_wait_set(timeout);
+  // Execute all the ready subscriptions
+  for (size_t i = 0; i < wait_set_.size_of_subscriptions; ++i) {
+    if (wait_set_.size_of_subscriptions && i < exec_list.number_of_subscriptions) {
+      if (wait_set_.subscriptions[i]) {
+        if (exec_list.subscription[i]->get_intra_process_subscription_handle()) {
+          execute_intra_process_subscription(exec_list.subscription[i]);
+        } else {
+          execute_subscription(exec_list.subscription[i]);
+        }
+      }
+    }
+  }
+  // Execute all the ready timers
+  for (size_t i = 0; i < wait_set_.size_of_timers; ++i) {
+    if (wait_set_.size_of_timers && i < exec_list.number_of_timers) {
+      if (wait_set_.timers[i] && exec_list.timer[i]->is_ready()) {
+          execute_timer(exec_list.timer[i]);
+      }
+    }
+  }
+  // Execute all the ready services
+  for (size_t i = 0; i < wait_set_.size_of_services; ++i) {
+    if (wait_set_.size_of_services && i < exec_list.number_of_services) {
+      if (wait_set_.services[i]) {
+          execute_service(exec_list.service[i]);
+      }
+    }
+  }
+  // Execute all the ready clients
+  for (size_t i = 0; i < wait_set_.size_of_clients; ++i) {
+    if (wait_set_.size_of_clients && i < exec_list.number_of_clients) {
+      if (wait_set_.clients[i]) {
+          execute_client(exec_list.client[i]);
+      }
+    }
+  }
+  // Execute all the ready waitables
+  for (size_t i = 0; i < exec_list.number_of_waitables; ++i) {
+    if (exec_list.number_of_waitables && exec_list.waitable[i]->is_ready(&wait_set_)) {
+      exec_list.waitable[i]->execute();
+    }
+  }
+  // Check the guard_conditions to see if anything is added to the executor
+  for (size_t i = 0; i < wait_set_.size_of_guard_conditions; ++i) {
+    if (wait_set_.guard_conditions[i]) {
+      // rebuild the wait_set and ExecutableList struct
+      run_collect_entities();
```

Comment:
At first we implemented the executor to be fully static, but this was quite restrictive. Now we use the node guard_conditions to check if something was added to a node at runtime. If this happens then we rebuild. This means that if nothing is added at runtime (everything is in place before .spin() ), this executor still behaves as a fully static one. So this check and rebuilding step only adds functionality. This means the executor can now be used for a wider variety of source code and will most likely still perform better than the original executor, since only rebuilding when necessary to me seems better than always rebuilding.

Reply:
@fujitatomoya 
> can we just recreate entities and executable list via user's request for reconstruction? i believe that it is easier to implement, and besides this user event or request for reconstruction would not be happening every single time just like SingleThreadExecutor does. so i think that it is reasonable.

Instead of rebuilding on request, we handled it like this (event trigger from node).

Reply:
i think that your idea is better than giving the responsibility to user for reconstruction.

Reply:
i believe that this reconstruction only needs to be done here once, so find out if anything is notified, and then reconstruct just once and exit from this loop. this is performance concern if some of entities are updated.

Reply:
Yes this is true, I believe we did realize this, but apparently haven't updated it here. We can add a "break;" after get_executable_list, that should get the desired result right?

Reply:
A guard condition is for "waking up" the executor, and I think that shouldn't be coupled to re-collecting entities (I'm not sure if you always want to re-collect entities when a guard condition is triggered).
IMO, it's better to provide a way of doing it under user request.

Reply:
One of the problems that we found is that this condition causes `run_collect_entities` to be called every time a guard condition is triggered, even if no nodes/pubs/subs/etc are added